File: mac/Tests/ClaudeToolbarCoreTests/Mascot/ClawdSpriteTests.swift
import XCTest
@testable import ClaudeToolbarCore

final class ClawdSpriteTests: XCTestCase {
    func testPoseFollowsTheArmAngle() {
        let cases: [(Double, ClawdPose)] = [(-60, .rest), (-30, .rest), (-29.9, .mid), (10, .mid), (10.1, .up), (40, .up), (40.1, .high), (75, .high)]
        for (angle, expected) in cases {
            XCTAssertEqual(ClawdSprite.poseFor(armAngle: angle), expected, "angle \(angle)")
        }
    }

    func testRestSpriteHasTheDecodedShape() {
        let cells = ClawdSprite.cells(for: .rest)
        XCTAssertEqual(cells.count, 56)
        XCTAssertEqual(Set(cells).count, 56)
        XCTAssertTrue(cells.allSatisfy { (0..<ClawdSprite.columns).contains($0.col) && (0..<ClawdSprite.rows).contains($0.row) })
        XCTAssertEqual(cells.filter { $0.kind == .eye }.sorted { $0.col < $1.col }, [SpriteCell(col: 5, row: 2, kind: .eye), SpriteCell(col: 12, row: 2, kind: .eye)])
        XCTAssertTrue(cells.contains(SpriteCell(col: 1, row: 3, kind: .body)))
        XCTAssertTrue(cells.contains(SpriteCell(col: 2, row: 3, kind: .body)))
        XCTAssertEqual(cells.filter { $0.row == 5 }.map(\.col).sorted(), [4, 6, 11, 13])
        XCTAssertFalse(cells.contains { $0.row == 0 })
        XCTAssertEqual(cells.filter { $0.row == 1 }.count, 12)
    }

    func testRightArmMovesWithThePose() {
        let expected: [ClawdPose: [SpriteCell]] = [
            .rest: [SpriteCell(col: 15, row: 3, kind: .body), SpriteCell(col: 16, row: 3, kind: .body)],
            .mid: [SpriteCell(col: 15, row: 3, kind: .body), SpriteCell(col: 16, row: 2, kind: .body)],
            .up: [SpriteCell(col: 15, row: 2, kind: .body), SpriteCell(col: 16, row: 1, kind: .body)],
            .high: [SpriteCell(col: 15, row: 1, kind: .body), SpriteCell(col: 16, row: 0, kind: .body)],
        ]
        for (pose, arm) in expected {
            let cells = ClawdSprite.cells(for: pose)
            XCTAssertEqual(cells.filter { $0.col >= 15 }.sorted { $0.col < $1.col }, arm, "\(pose)")
            XCTAssertEqual(cells.count, 56)
        }
    }

    func testColoursAndBadgeGeometryAreFixed() {
        XCTAssertEqual(ClawdSprite.bodyColor, "#FFD97757")
        XCTAssertEqual(ClawdSprite.eyeColor, "#FF1B1B1B")
        XCTAssertEqual(ClawdSprite.badgeCenterCol, 16.5)
        XCTAssertEqual(ClawdSprite.badgeCenterRow, 0.6)
        XCTAssertEqual(ClawdSprite.badgeRadiusCells, 2)
        XCTAssertNil(MascotBadge.none.hex)
        XCTAssertEqual(MascotBadge.attention.hex, "#FFD29922")
        XCTAssertEqual(MascotBadge.working.hex, "#FF3B82F6")
        XCTAssertEqual(MascotBadge.finished.hex, "#FF3FB950")
        XCTAssertEqual(MascotBadge.failed.hex, "#FFF85149")
        XCTAssertEqual([MascotBadge.working, .attention, .failed, .finished].max(), .attention)
    }
}

File: tests/ClaudeToolbar.Core.Tests/Mascot/ClawdSpriteTests.cs
using ClaudeToolbar.Core.Mascot;

namespace ClaudeToolbar.Core.Tests.Mascot;

public class ClawdSpriteTests
{
    [Theory]
    [InlineData(-60, ClawdPose.Rest)]
    [InlineData(-30, ClawdPose.Rest)]
    [InlineData(-29.9, ClawdPose.Mid)]
    [InlineData(10, ClawdPose.Mid)]
    [InlineData(10.1, ClawdPose.Up)]
    [InlineData(40, ClawdPose.Up)]
    [InlineData(40.1, ClawdPose.High)]
    [InlineData(75, ClawdPose.High)]
    public void PoseFollowsTheArmAngle(double angle, ClawdPose expected) => Assert.Equal(expected, ClawdSprite.PoseFor(angle));

    [Fact]
    public void RestSpriteHasTheDecodedShape()
    {
        var cells = ClawdSprite.Cells(ClawdPose.Rest);
        Assert.Equal(56, cells.Count);
        Assert.Equal(cells.Count, cells.Distinct().Count());
        Assert.All(cells, c => Assert.InRange(c.Col, 0, ClawdSprite.Columns - 1));
        Assert.All(cells, c => Assert.InRange(c.Row, 0, ClawdSprite.Rows - 1));
        Assert.Equal([new SpriteCell(5, 2, CellKind.Eye), new SpriteCell(12, 2, CellKind.Eye)], cells.Where(c => c.Kind == CellKind.Eye).OrderBy(c => c.Col).ToList());
        Assert.Contains(new SpriteCell(1, 3, CellKind.Body), cells);
        Assert.Contains(new SpriteCell(2, 3, CellKind.Body), cells);
        Assert.Equal([4, 6, 11, 13], cells.Where(c => c.Row == 5).Select(c => c.Col).OrderBy(c => c).ToList());
        Assert.DoesNotContain(cells, c => c.Row == 0);
        Assert.Equal(12, cells.Count(c => c.Row == 1));
    }

    [Theory]
    [InlineData(ClawdPose.Rest, 15, 3, 16, 3)]
    [InlineData(ClawdPose.Mid, 15, 3, 16, 2)]
    [InlineData(ClawdPose.Up, 15, 2, 16, 1)]
    [InlineData(ClawdPose.High, 15, 1, 16, 0)]
    public void RightArmMovesWithThePose(ClawdPose pose, int c1, int r1, int c2, int r2)
    {
        var arm = ClawdSprite.Cells(pose).Where(c => c.Col >= 15).OrderBy(c => c.Col).ToList();
        Assert.Equal([new SpriteCell(c1, r1, CellKind.Body), new SpriteCell(c2, r2, CellKind.Body)], arm);
        Assert.Equal(56, ClawdSprite.Cells(pose).Count);
    }

    [Fact]
    public void ColoursAndBadgeGeometryAreFixed()
    {
        Assert.Equal("#FFD97757", ClawdSprite.BodyColor);
        Assert.Equal("#FF1B1B1B", ClawdSprite.EyeColor);
        Assert.Equal(16.5, ClawdSprite.BadgeCenterCol);
        Assert.Equal(0.6, ClawdSprite.BadgeCenterRow);
        Assert.Equal(2, ClawdSprite.BadgeRadiusCells);
        Assert.Null(MascotBadgeColors.Hex(MascotBadge.None));
        Assert.Equal("#FFD29922", MascotBadgeColors.Hex(MascotBadge.Attention));
        Assert.Equal("#FF3B82F6", MascotBadgeColors.Hex(MascotBadge.Working));
        Assert.Equal("#FF3FB950", MascotBadgeColors.Hex(MascotBadge.Finished));
        Assert.Equal("#FFF85149", MascotBadgeColors.Hex(MascotBadge.Failed));
    }
}
